src/libraries/System.Data.Common/src/System/Xml/XPathNodePointer.cs

```diff
@@ -433,20 +431,20 @@ internal string XmlLang
                         }
                     }
                     if (curNode.NodeType == XmlNodeType.Attribute)
-                        curNode = ((XmlAttribute)curNode).OwnerElement;
+                        curNode = ((XmlAttribute)curNode).OwnerElement!;
                     else
-                        curNode = curNode.ParentNode;
+                        curNode = curNode.ParentNode!;
                 }
                 return string.Empty;
             }
         }
 
-        private XmlBoundElement GetRowElement()
+        private XmlBoundElement? GetRowElement()
         {
-            XmlBoundElement rowElem;
+            XmlBoundElement? rowElem;
             if (_column != null)
             {
-                rowElem = _node as XmlBoundElement;
+                rowElem = (XmlBoundElement)_node;
```

Comment:
I mean, technically the change is not _entirely_ wrong (aside from the fact that now it could throw if the cast fails), but it's not a change related to nullability.
Up to you if you want to revert it.

Reply:
If the assertions are wrong, this still represents a change of behavior from returning null to throwing InvalidCastException plus the method is already expected to return null.

Reply:
The previous code was this:

```
rowElem = _node as XmlBoundElement;
Debug.Assert(rowElem != null);
```

* If the cast doesn't succeed, the next line asserts
* If the _node is null, the cast will succeed but the next line asserts
* In release builds, a null reference would be returned

My reasoning was: a hard cast seems more desirable because at least that makes the code fail consistently, rather than asserting in debug builds an null referencing elsewhere in production.

But if you guys feel we shouldn't do this, I can revert.